packages/core/src/browser/tree/tree-selection.ts

```diff
@@ -5,56 +5,112 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { injectable, inject } from "inversify";
-import { Event, Emitter, Disposable, SelectionProvider } from "../../common";
-import { ITree, ITreeNode } from "./tree";
-
-export const ITreeSelectionService = Symbol("ITreeSelectionService");
+import { injectable, inject, postConstruct } from 'inversify';
+import { Tree, TreeNode } from './tree';
+import { TreeSelectionState } from './tree-selection-state';
+import { Event, Emitter, Disposable, SelectionProvider } from '../../common';
 
 /**
  * The tree selection service.
  */
-export interface ITreeSelectionService extends Disposable, SelectionProvider<Readonly<ISelectableTreeNode> | undefined> {
+export const TreeSelectionService = Symbol("TreeSelectionService");
+export interface TreeSelectionService extends Disposable, SelectionProvider<ReadonlyArray<Readonly<SelectableTreeNode>>> {
+
+    /**
+     * The tree selection, representing the selected nodes from the tree. If nothing is selected, the
+     * result will be empty.
+     */
+    readonly selectedNodes: ReadonlyArray<Readonly<SelectableTreeNode>>;
+
+    /**
+     * Emitted when the selection has changed in the tree.
+     */
+    readonly onSelectionChanged: Event<ReadonlyArray<Readonly<SelectableTreeNode>>>;
+
+    /**
+     * Registers the given selection into the tree selection service. If the selection state changes after adding the
+     * `selectionOrTreeNode` argument, a selection changed event will be fired. If the argument is a tree node,
+     * a it will be treated as a tree selection with the default selection type.
+     */
+    addSelection(selectionOrTreeNode: TreeSelection | Readonly<SelectableTreeNode>): void;
+
+}
+
+/**
+ * Representation of a tree selection.
+ */
+export interface TreeSelection {
+
     /**
-     * The node selected in the tree. If defined then valid.
-     * Undefined if there is no node selection.
+     * The actual item that has been selected.
      */
-    readonly selectedNode: Readonly<ISelectableTreeNode> | undefined;
+    readonly node: Readonly<SelectableTreeNode>;
+
     /**
-     * Emit when the node selection is changed.
+     * The optional tree selection type. Defaults to `SelectionType.DEFAULT`;
      */
-    readonly onSelectionChanged: Event<Readonly<ISelectableTreeNode> | undefined>;
+    readonly type?: TreeSelection.SelectionType;
+
+}
+
+export namespace TreeSelection {
+
     /**
-     * Select a given node.
-     * If a given node is undefined or invalid then remove the node selection.
+     * Enumeration of selection types.
      */
-    selectNode(node: Readonly<ISelectableTreeNode> | undefined): void;
+    export enum SelectionType {
+
+        /**
+         * When selecting a tree node without any key masks.
```

Comment:
The tree widget is a client and a reflection of the tree. The tree does care about UI and key masks.

Reply:
What is your actual concern, to documentation or what exactly?

Reply:
Yes, it should not mention key masks, it is implementation details of the client.